### PR TITLE
generalise the sandbox

### PR DESCRIPTION
**Pull request by patricoferris**
This is in partial fulfilment of #57 -- in particular making the `S.SANDBOX` interface more general by removing runc/linux specific from `build.ml`. 

Currently OBuilder supports Linux by using [runc](https://github.com/ocurrent/obuilder/blob/6438c3ee7530bd6d73234b4b39360ac4af9d56d3/lib/runc_sandbox.ml#L275) and a little bit of [docker](https://github.com/ocurrent/obuilder/blob/6438c3ee7530bd6d73234b4b39360ac4af9d56d3/lib/build.ml#L250). The docker component is currently in the more agnostic `lib/build.ml` side of things, and the `get_base` function is the linux-y interpretation of `(from <hash>)`.  For MacOS support (and perhaps other platforms) the interpretation of `(from <hash>)` will be different and is therefore dependent on the sandboxing environment -- I think it makes sense to force the sandboxing environment to provide this functionality. To this end, this PR: 

 - Extended the interface for sandboxes by providing:
   + `from` which is used where the original `get_base` function was used -- this is a function that is passed to `Store.build`.
   + `create` before this was an additional function added to `runc_sandbox.mli` but MacOS will also require a create function and since the choice of sandboxing environment will likely be a compile-time choice, it makes sense I think to unify this `create` function type by (a) putting it in `S.SANDBOX` and (b) extracting a `config` type that has a `cmdliner` value that can be used to build a new `S.SANDBOX.t`
   + The create function has an optional `state_dir` parameter -- the reason this isn't baked into the `config` of the `runc` sandbox is that I think it is often provided at runtime, e.g. https://github.com/ocurrent/ocluster/blob/master/worker/obuilder_build.ml#L37

### CODE REVIEW

File: lib/runc_sandbox.ml
Comment (avsm) [inline lib/runc_sandbox.ml:386]:
```suggestion
    ~doc:"Install a seccomp filter that skips all synchronous syscalls"
```

File: lib/s.ml
Comment (avsm) [inline lib/s.ml:67]: can this just be a `val pp : t Fmt.t` instead? That would let you pretty print the abstract sandbox state, and it could simply prepend the sandbox type there as well.

Comment (patricoferris) [reply lib/s.ml:67]: Sound very reasonable to me, thanks :)) Will also update the doc comment from above shortly

Comment (patricoferris) [reply lib/s.ml]: It looks something like:
```
runc state:
 {runc_state_dir: /hello/world
  fast_sync: true
  arches: [A,B,C]}
```

Hopefully that's okay, I'm no `Format` wizard 🧙 .

File: lib/runc_sandbox.ml
Comment (talex5) [reply lib/runc_sandbox.ml]: But the new text is wrong! The option skips all `sync` and related syscalls, not all synchronous syscalls.

File: lib/build.ml
Comment (talex5) [inline lib/build.ml:246]: We could just keep this bit of `get_base`, I think, instead of making every sandbox duplicate it.

File: lib/build.ml
Comment (talex5) [inline lib/build.ml:263]: ... and keep this bit too.

File: lib/build.ml
Comment (talex5) [inline lib/build.ml:310]: If you keep `get_base`, this part doesn't need to be duplicated.

File: lib/s.ml
Comment (talex5) [reply lib/s.ml:77]: Also, not sure that `config` and `create` should be in `SANDBOX`. I see this already caused some trouble for the mock sandbox. This could stay in `runc_sandbox.mli` for now, and then when there are multiple implementations, we can have this in a shared `sandbox.mli`.

**Review by talex5 — COMMENTED** [review]
This looks like generally the right direction - sorry for all the nit-picking!

File: test/mock_sandbox.ml
Comment (patricoferris) [reply test/mock_sandbox.ml:17]: Yep

File: test/mock_sandbox.ml
Comment (patricoferris) [reply test/mock_sandbox.ml:104]: Okay I think I follow -- would it be enough then for the `mock_sandbox` to just be: 

```ocaml
let from : Obuilder.Runc_sandbox.from
```

for now?

**Review by patricoferris — COMMENTED** [review]
Thanks for the detailed feedback, have responded to various bits but I think these changes clean the code up quite a bit!

File: test/mock_sandbox.ml
Comment (talex5) [reply test/mock_sandbox.ml:104]: That would be an easy solution for now, yes.

Probably the `from` stuff should go in a separate module anyway (so we have three concepts: stores, sandboxes, and image sources). It would be quite possible to use `docker pull` even on macos, for example.

File: lib/build.ml
Comment (talex5) [reply lib/build.ml:263]: Yes. You could also create `rootfs` here too, and pass that to the fetcher, so it doesn't need to know about the layout.

Comment (patricoferris) [reply lib/build.ml:263]: I think I may have been mistaken about the implementation. Am I right in thinking that the runc/zfs implementation builds inside of `/tank/result/<hash>/rootfs/` ? So that's where I should build the MacOS one too? Up until now I've just been building it inside of `/tank/result/<hash>`, to make it uniform I can use `rootfs` which as you said reduces the amount of things the fetcher needs to do/know about?

Comment (talex5) [reply lib/build.ml:263]: Yes, that's right. It allows us to store extra data about the build, such as the environment variables and the log, plus anything else we think of in the future.

**Comment by talex5** [comment]
What's happening with this? Is there a new version available to review?

I'd like to make some improvements to the docs, but I don't want to create conflicts with this work.

**Comment by patricoferris** [comment]
Hi @talex5 :)) I would say conflict away, this is currently somewhat shelved in favour me testing the idea more (e.g. deploying a little version of opam-health-check with my own modified version of OCluster/OBuilder). Sorry for not mentioning that.

**Comment by talex5** [comment]
I've updated the docs, so this needs rebasing now.

It would be good to get this in soon, because we're going to need it for the Windows support too.

**Comment by patricoferris** [comment]
Okay 👍  -- will work on it this afternoon

**Comment by patricoferris** [comment]
@talex5 this should be ready for a second review, sorry about the weird formatting errors I had to fix. I still have a few questions though.

Re:https://github.com/ocurrent/obuilder/pull/58#discussion_r554558472 do you want this in this PR? I took this implementation and used it with [ocluster to see what it would look like](https://github.com/ocurrent/ocluster/compare/master...patricoferris:generalise-sandbox) and because `cmdliner/config` is in `Runc_sandbox.mli` everything is... not very generalised 😅 ?

After re-writing everything to go with your suggestions, this comment https://github.com/ocurrent/obuilder/pull/58#discussion_r555014886 really resonated with me. Do you think there should be a `S.FETCH` with a `docker.ml` implementation or something like that (if so in this PR or a different one?).

Lastly I haven't passed in `tmp / "rootfs"` -- I wasn't 100% sure all implementations would use that directory so just passing `tmp` and forcing the implementation to decide seemed safest, but let me know if that's not the case :)

**Comment by talex5** [comment]
Thanks, this is good progress :-)

> Re:#58 (comment) do you want this in this PR? I took this implementation and used it with ocluster to see what it would look like and because cmdliner/config is in Runc_sandbox.mli everything is... not very generalised?

Well, the interface should eventually be in `sandbox.mli`, not `runc_sandbox.mli`, and then it should look OK. If you prefer, I don't mind creating another signature for this (e.g. `STORE_MAKER` or something), as long as it's separate from `STORE`.

> Do you think there should be a S.FETCH with a docker.ml implementation or something like that (if so in this PR or a different one?).

I think that makes sense, especially as the Windows implementation will be using `docker pull` too. For now, I'd suggest just putting the docker-pull stuff in its own module (since you're moving it anyway).

> Lastly I haven't passed in tmp / "rootfs" -- I wasn't 100% sure all implementations would use that directory so just passing tmp and forcing the implementation to decide seemed safest, but let me know if that's not the case :)

It's not a good idea to let the sandboxes/fetchers pick a name themselves because:

1. It has to not conflict with the names used in the main library (e.g. `log`), and we may add new ones later.
2. It has to be the same for fetchers and sandboxes.

**Comment by MisterDA** [comment]
Nothing much really, but you've left trailing whitespace in the obuilder PR.
Could you do a little run of something akin to this?
`sed -i'' 's/[[:space:]]*$//' **/*.ml{i,}`

Passing the `conf` parameter doesn't play nicely with multiple sandboxes implementation: I have the `Runc_sandbox.config` type and the `Docker_sandbox.config` that conflict. I'm thinking of moving it to the `S.SANDBOX` module, a bit like the `BUILDER.context` type.

**Comment by MisterDA** [comment]
Thanks for the explanation!
The sandbox doesn't have to become a build choice, it can be selected according to the chosen storage. In the case of the Docker sandbox, it makes sense to decide to use it depending on the storage: if the Docker storage is used, then the Docker sandbox should be used too, otherwise (btrfs/zfs) use runc. On Windows, if (when) we'll use a snapshotting filesystem, we'll also switch to the port of runc.

**Comment by patricoferris** [comment]
Would be good to get this merged, as it is now blocking @MisterDA. 

To summarise the changes: 
 - There is now a notion of `FETCHER` to perform the initialisation of the `rootfs` directory. For now there is one implementation, `Docker`, but macOS currently does something completely different so it is nice to have this flexibility.
 - Sandboxes have been generalised to be passed a `config` argument which is implementation specific and a `config Cmdliner.Term.t` value is provided so the CLI(s) can expose the configuration parameters, in runc's case this is just the `fast_sync` functionality. At the moment I don't think we're using OBuilder anywhere where we can't use a CLI so the fact that the config is abstract and can only be built via the CLI is ok, but I'm not 100% sure on this.
 - The build log copy function which is used in multiple places now lives in `build_log.ml`
 
I think that's everything, apart from some formatting what do you think in terms of next steps for this @talex5? @MisterDA too ? Sorry it's taken so long!

**Review by talex5 — APPROVED** [review]
Thanks! Looks good to merge once CI is done (I just rebased it and reindented the `get_base` function).

**Pull request merged**
